[17기 박재현] 리뷰 좋아요 뷰 작성

namujinju opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 리뷰 좋아요 뷰 작성
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- product 앱의 ReviewLikeView 생성
- ProductReviewView에서 order 값이 없으면 최신순으로 default 값 변경

## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

namujinju committed:
Add: 리뷰 좋아요 뷰 작성
- product 앱의 ReviewLikeView 생성
- ProductReviewView에서 order 값이 없으면 최신순으로 default 값 변경

soheon-lee commented on product/views.py:
```diff
@@ -51,5 +52,48 @@ def get(self, request, product_id):
+        try:
+            data     = json.loads(request.body)
+            review_id = data.get('review_id')
+
+            # user는 login_decorator로 불러온다.
+            email    = data.get('email')
+            password = data.get('password')
+
+            if not email:
+                return JsonResponse({'message':'NO_EMAIL_ERROR'}, status=400)
+
+            if not User.objects.filter(email=email).exists():
+                return JsonResponse({'message':'INVALID_USER'}, status=401)
+
+            user = User.objects.get(email=email)
+
```
이 부분은 `login decorator`에서 해주어야 하는 부분입니다.
에러처리 또한 데코레이터 내부에서만 해주시면 됩니다 :)

namujinju replied:
네. 아직 login decorator가 완성되지 않아서 통신을 위해 임시로 만든 부분입니다. decorator 적용해서 다시 push 하겠습니다.

soheon-lee commented on product/views.py:
```diff
@@ -51,5 +52,48 @@ def get(self, request, product_id):
+
+            user = User.objects.get(email=email)
+
+            if ProductReview.objects.filter(id=review_id).exists():
+                product_review = ProductReview.objects.get(id=review_id)
+            else:
+                return JsonResponse({'message':'INVALID_REVIEW'}, status=401)
+            
```
```suggestion
            if not ProductReview.objects.filter(id=review_id).exists():
                return JsonResponse({'message':'INVALID_REVIEW'}, status=401)

            product_review = ProductReview.objects.get(id=review_id)
```
조건이 충족되지 않았을 때에 함수가 바로 return 할 수 있도록 하시면 더 좋을 것 같아요 👍

namujinju replied:
네 수정하겠습니다!

soheon-lee commented on product/views.py:
```diff
@@ -51,5 +52,48 @@ def get(self, request, product_id):
+            
+            if product_review.user == user:
+                return JsonResponse({'message':'CANNOT_LIKE_YOUR_REVIEW'}, status=401)
+            
+            if ReviewLike.objects.filter(review=product_review, user=user).exists():
+                ReviewLike.objects.filter(review=product_review, user=user).delete()
+            else:
+                ReviewLike.objects.create(review=product_review, user=user)
+        
```
이 조건을 만족하는 `ReviewLike`클래스가 있으면 삭제하고, 없으면 생성하는 로직인 것이죠? 좋습니다. 👍 👍 👍 
이와 비슷한 로직으로 장고에서는 데이터가 있으면 가져오고, 없으면 생성해주는 메소드가 있습니다. (지금과 같은 경우라면 데이터가 있을 때 가져와서 삭제하면 되겠죠?) 한 번 찾으셔서 적용해보시면 코드가 좀 더 간결해질 것 같네요!

namujinju replied:
update_or_create와 원리가 비슷한 메소드인 것 같네요. 잘 찾아보겠습니다!

namujinju replied:
말씀하시는 함수를 찾아서 적용했습니다. 감사합니다!

soheon-lee commented on product/views.py:
```diff
@@ -51,5 +52,48 @@ def get(self, request, product_id):
+            if product_review.user == user:
+                return JsonResponse({'message':'CANNOT_LIKE_YOUR_REVIEW'}, status=401)
+            
+            if ReviewLike.objects.filter(review=product_review, user=user).exists():
```
user 인스턴스는 ReviewLike 테이블과 연결되어 있으므로,
아예 새로 ReviewLike 클래스에 접근하시는 방법 외에도 user 인스턴스에서 시작하는 방법으로 한 번 응용해보세요 :)
`user.______.filter().exists()`로 시작해보시죠 ! 화이팅

namujinju replied:
네 감사합니다!

namujinju replied:
아직도 굉장히 헷갈리는 부분인 것 같습니다. 다시 공부해서 블로깅으로 잘 정리하겠습니다. 감사합니다.

namujinju committed:
Refactor: 리뷰 좋아요 뷰 지시 사항 반영
- login decorator 추가
- 조건이 충족되지 않았을 때 바로 return 하도록 수정
- 좋아요와 좋아요 취소를 get_or_create 함수를 이용해 간결하게 수정
- 테이블 간의 관계를 고려한 인스턴스 접근 방법 변경

namujinju committed:
Delete: migration 파일 삭제

namujinju committed:
Merge branch 'main' into feature/reviewlike

namujinju committed:
Delete: utils.py

namujinju committed:
Modify: add import login_decorator in product.views

soheon-lee commented:
LGTM!
